File: make.js
const fs = require('fs')
const glob = require('glob')
const mkdirp = require('mkdirp')
//const stencila = require('stencila')
const xmldom = require('xmldom')
const xpath = require('xpath.js')
const zlib = require('zlib')

mkdirp('build', function (err) {
  if (err) throw err

  // For checking that the same function name is not used twice
  let names = {}

  // The compiled XML function definitions
  let vfs = fs.createWriteStream('build/funcs.js')
  vfs.write('window.STENCILA_LIB_FUNCTIONS = {\n')

  // The compiled module of JavaScript Function implementations
  let lib = fs.createWriteStream('build/lib.js')

  // For each function...
  glob('**/*.fun.xml', function (err, files) {
    if (err) throw err

    files.forEach(function(path) {
      if (path.substring(0,1) === '.') return

      console.info(path)
      fs.readFile(path, 'utf8', function(err, xml) {
        if (err) throw err

        const doc = new xmldom.DOMParser().parseFromString(xml)
        
        // Check that the function name has not already been used
        const name = xpath(doc, '/function/name/text()')[0].data
        if (names[name]) throw new Error(`Function name already used: "${name}"`)
        else names[name] = true

        // Load the document into memory
        //const func = stencila.importFunction(xml)
          
        // Extract any Javascript implementations for the compiled module
        xpath(doc, '/function/implems/implem[@language="js"]').forEach(function (implem) {
          const types = xpath(implem, './types/type/@type')
          const signat = (['mini', name].concat(types)).join('_')
          const code = xpath(implem, './code/text()')[0].data

          lib.write(`var ${signat} = (${code})\n\n`)
        })
        
        // Write a compacted version of XML to the definitions file
        const xmlCompact = xml.replace(/>\s*</g,'><').replace(/'/g,"\\'").trim()
        vfs.write(`  ${name}: '${xmlCompact}',\n`)
      })
    })
  })

  process.on('exit', function() {
    vfs.write('}\n')
  })

})

File: py/make.py
import glob
import lxml.etree
import os
import re

# For checking that the same function name is not used twice
names = []

# The compiled Python module of function implementations
if not os.path.exists('build/py'):
    os.makedirs('build/py')
lib = open('build/py/lib.py', 'w')
lib.write('''# Modules available for functions
import math
import numpy
import types

''')

# Function implementation type regex
def_regex = re.compile(r'^(def\s+\w+)\s*\(')

# For each function...
for path in glob.glob('**/*.fun.xml'):
    print path
    with open(path) as file:
        doc = lxml.etree.parse(file)

        # Check that the function name has not already been used
        name = doc.xpath('/function/name/text()')[0]
        if name in names:
            raise RuntimeError('Function name alredy used: ' + name)
        names.append(name)

        # Get parameter types for the signature
        param_types = doc.xpath('/function/params/param/@type')

        # Extract any Python implementations and put them in the Python package
        for implem in doc.xpath('/function/implems/implem[@language="py"]'):
            types = implem.xpath('./types/type/@type')
            if len(types) == 0:
                types = param_types
            signat = '_'.join([name] + types).replace('[', '_').replace(']', '')
            code = implem.xpath('./code/text()')[0]

            match = def_regex.match(code)
            if match:
                defn = match.group(1)
                lib.write(code.replace(defn, 'def %s' % name) + '\n\n')
            else:
                lib.write('%s = %s\n\n\n' % (name, code))
